ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -420,14 +423,14 @@ shakeDatabaseProfileIO mbProfileDir = do
                 return (dir </> file)
 
 setValues :: IdeRule k v
-          => Var Values
+          => Values
           -> k
           -> NormalizedFilePath
           -> Value v
           -> Vector FileDiagnostic
           -> IO ()
 setValues state key file val diags =
-    void $ modifyVar' state $ HMap.insert (toKey key file) (ValueWithDiagnostics (fmap toDyn val) diags)
+    atomically $ STM.insert (ValueWithDiagnostics (fmap toDyn val) diags) (toKey key file) state
```

Comment:
I wonder whether for things that are in `IO` and end with a top-level `atomically` it would make sense to change the type to stay in STM so the caller has the option of batching it if they want to?

Reply:
I agree, it makes sense to move the `atomically` to the call site in these cases